### PR TITLE
Reduxify: lib/username

### LINKED ISSUE #24128: Reduxify: lib/username
## Summary
A **small** emitter that works with /me to change a user's username. No actions, no flux. Might be a good candidate for @jsnajdr's "new" data approach.

### Code
[lib/username](https://github.com/Automattic/wp-calypso/tree/master/client/lib/username)

### Imports
Only used by client/me/account/controller.js. 

### Usage
Passed down to `me/account/main` where's it's observed using the `observe` mixin. Methods are called directly on it from the component.

### PR DESCRIPTION
**Pull request by jsliang**
Trying to fix #24128.
This PR is also part of #24004 as it removes `lib/user` from `lib/username`.
- Reduxifies `lib/username`
- Uses a query component to send network requests for username validation
- Decouples `lib/user` from `lib/username` by moving the relevant logic to `me/account/main.jsx`
- Adds unit test cases

**To Test**
- Open http://calypso.localhost:3000/me/account
- Type 3 characters in the **Username** input box and confirm that an error message `Usernames must be at least 4 characters.` appears
- Type random characters with invalid characters (i.e. not lowercase letters and not numbers) in the **Username** input box and confirm that an error message `
Usernames can only contain lowercase letters (a-z) and numbers.` appears
- Type an existing username in the **Username** input box and confirm that an error message `Sorry, that username already exists!` appears
- Type a valid and non-existing username in the **Username** input box and confirm that a message `<username> is a valid username.` appears
- Click the **Save Username** button at the bottom of the form and confirm that the username is changed after page reload

### CODE REVIEW

**Comment by designsimply** [comment]
> I would like to add a `[Status] In Progress` label to this PR, but I don't have write access.

I updated the label for you. I will try to keep an eye out for comments from you to move it to `[Status] Needs Review` when you're ready for that. 🙂 Please ping me directly if I don't respond within a day. I try to check everything daily but I get a large number of notifications on GitHub!

**Comment by jsliang** [comment]
@designsimply , thank you for adding the label. This PR is ready for review now. I've updated the PR comment describing what it does and how to test it. Please move it to `[Status] Needs Review` when you have the time. Thanks again in advance!

**Comment by designsimply** [comment]
Done! Moved to `[Status] Needs Review`.

File: client/components/data/query-username-validation/index.jsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+		this.props.validateUsername( username );
+	}
+
+	render() {
+		return null;
+	}
+}
+
+export default connect( null, dispatch =>
+	bindActionCreators(
```
Comment (blowery) [inline client/components/data/query-username-validation/index.jsx]: There's a handy shorthand here. You can just pass 
```
{
  clearUsernameValidation,
  validateUsername,
}
```
and it'll auto bind them to `dispatch`

File: client/state/username/reducer.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * Internal dependencies
+ */
+import {
+	USERNAME_CLEAR_VALIDATION,
+	USERNAME_VALIDATION_FAILURE,
+	USERNAME_VALIDATION_SUCCESS,
+} from 'state/action-types';
+import { USERNAME_DEFAULT } from './constants';
+
+export default function( state, action = {} ) {
```
[the remark this answers is not in the document]
Comment (blowery) [reply client/state/username/reducer.js]: though in this case, an empty object or `undefined` is probably a fine default

**Comment by jsliang** [comment]
Hi @blowery,
Thank you for all the comments! Yes, I'm still interested in this PR. It's been a busy week for me, but I believe I'll be able to make changes this weekend.

File: client/components/data/query-username-validation/index.jsx [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+		if ( this.props.username !== prevProps.username ) {
+			this.debouncedValidateUsername();
+		}
+	}
+
+	componentWillUnmount() {
+		this.props.clearUsernameValidation();
+	}
+
+	debouncedValidateUsername = debounce( this.requestForValidation, 600 );
```
Comment (dmsnell) [inline client/components/data/query-username-validation/index.jsx]: what's the purpose for the debouncing here?

**Comment by dmsnell** [comment]
thanks for the poke @blowery - yeah this is a perfect candidate for `httpData` and we should be able to remove about six hundred lines of code or so 😄 

@jsliang we can talk about `httpData()` if you are interested

**Comment by jsliang** [comment]
Hi @dmsnell,
Yes, I'm interested in knowing more about `httpData`. Is there any discussion or documentation that I can refer?

**Comment by blowery** [comment]
@jsliang there are some examples in https://github.com/Automattic/wp-calypso/blob/master/client/state/data-getters/index.js with an example usage at https://github.com/Automattic/wp-calypso/blob/master/client/me/account/main.jsx#L818

**Comment by dmsnell** [comment]
Thanks for your patience @jsliang!

The places we would want to start I think aren't actually in the code itself or in the `httpData` library but in asking what the purpose is of this lib, how many callers use it, and how it's used.

It appears to be used to make sure that a username is available and meets the requirements on which characters are allowed and what-not. Is it burdensome to tally that information?

The thing about `httpData` is that it gives us a choice between "one-off" data requests and recurring requests. It also allows us to be laser-focused on the data we want or the requests we want to make. That is, if we only have a single component using `lib/username` we can probably just get rid of it and make a direct request for the validation.

The system operates with a few essential elements and none of it is documented yet because it's in the early phase of testing and gathering API feedback.
 - every piece of data gets manually assigned a unique string id. if we want to consider username validity then we could imaging the id like this `username-validity-${ username }`. We have the freedom to choose the id but we want to make sure we don't pick something generic or we'll likely overlap with another use (for example, `username` would be a poor choice due to its wide semantic range).
 - every piece of data is fetched with a basic Redux action. in Calypso that means either a raw HTTP request or a request to the WordPress.com API. these are created using the `http` and `rawHttp` functions.
 - requests fail or succeed and they return data. we create a `fromApi` function to convert the response of a remote request into the local data formats that makes sense inside of Calypso. if we throw any exceptions during this transformation the request automatically fails into the error mode.
 - on success our `fromApi` function returns a list of pairs of ids and data. this lets us update multiple bits of state from a single API call. for the username case we'd probably just return a single pair

```js
const usernameData = username => requestHttpData(
	`username-validity-${ username }`,
	http( { method: 'GET', path: '/is/username/valid', query: { username } ),
	{
		// don't worry about this empty () - just make sure it's there
		fromApi: () => ( { is_valid } ) => [
			[ `username-validity-${ username } )`, !! is_valid ]
		]
	}
);
```

The `requestHttpData()` function will return a data structure representing remote-requestable data. The `data` property will have the value of the given id as it stands when called. It will update automatically after the API call returns. If you want to know if the request is in progress you can check the `status` property. Usually we add this into `mapStateToProps`

```js
const mapStateToProps = ( state, { username } ) => {
	const { data: isValid, status } = usernameData( username );

	return {
		isValid,
		isChecking: status === 'pending',
	};
}
```

Hope this isn't too much or too overwhelming! Don't be a stranger 😉

**Comment by blowery** [comment]
We just had another PR land that would serve as a good example of where to head with this: https://github.com/Automattic/wp-calypso/issues/26407

**Comment by blowery** [comment]
Moving this one back to In Progress

**Comment by jsliang** [comment]
Hi @blowery and @dmsnell,

Sorry for the late update.

I've partially modified the PR to see if I'm on the same page with you. Only username validation is migrated to use `requestHttpData` for now. I plan to proceed with username changing part later.

I encounter a problem regarding the retry mechanism when trying out `requestHttpData`, and I feel that I need some guidance.

The validation API `/me/username/validate/{username}` returns 200 if the username is valid, and 400 otherwise. When I tested with a valid username, everything looked good. But when I tested with an invalid username that receives a 400 response, the request kept retrying forever, and it took several seconds for the error message to finally appear in the state of data-layer. Thus the user experience has been quite miserable.

To reproduce:
1. Open http://calypso.localhost:3000/me/account
2. Type a valid and non-existing username in the Username input box. You'll see that a message `<username> is a valid username.` appears in no time.
3. Open the debug tool of your browser to monitor network requests.
4. Type an invalid or existing username in the Username input box. You'll find that the error message takes about 5 or more seconds to appear. You can also see that the validation request is retried infinitely via the debug tool.

My guess of the solution is to change the retry mechanism of my `http` (from `state/data-layer/wpcom-http/actions`) call, but I failed to find any documentation explaining how to do it. Could you suggest how I can specify not to retry when making the `http` call? Or do you think the cause of the infinite retries might be something else?

Thank you.

**Comment by flootr** [comment]
@jsliang what's the current state of this PR? Do you still want to tackle it? Also looping @jsnajdr in here who mentioned some work he's doing in this area.

**Comment by jsnajdr** [comment]
I have some work in progress related to `lib/users`, the store powering the People/Team screen.

This `lib/username` is fairly independent, and uses `lib/user` only to check the current user's username and mark it immediately as "taken", without even asking the server 🙂

**Comment by jsliang** [comment]
> @jsliang what's the current state of this PR? Do you still want to tackle it? Also looping @jsnajdr in here who mentioned some work he's doing in this area.

Hi @flootr, sorry for not updating for over 2 years. Also, apologies that I don't have the capacity to proceed with this PR at the moment. Would it be better if I close this PR for now?

**Comment by flootr** [comment]
@jsliang no worries! Thanks for your response :)

**Pull request closed without merging**